File: jarvis-os/apps/api/src/routes/agents.ts
import { Router, Request, Response } from "express";
import { z } from "zod";
import { prisma } from "@jarvis/database";
import { runAgentWithTools } from "../agents/orchestrator";
import { AppError } from "../middleware/errorHandler";
import { createSuccessResponse } from "@jarvis/shared";
import { chatRateLimiter } from "../middleware/rateLimiter";

const router = Router();

const AGENT_DEFINITIONS = [
  {
    id: "research",
    name: "Research Agent",
    type: "research",
    description: "Searches the web, gathers information, and builds reports",
    capabilities: ["web_search", "summarize", "fact_check", "report_generation"],
    tools: ["web_search", "search_memory", "save_memory"],
    model: "claude-sonnet-4-6",
  },
  {
    id: "coding",
    name: "Coding Agent",
    type: "coding",
    description: "Generates, analyzes, and debugs code across all languages",
    capabilities: ["code_generation", "debugging", "code_review", "git_operations"],
    tools: ["analyze_code", "web_search", "search_memory"],
    model: "claude-sonnet-4-6",
  },
  {
    id: "planning",
    name: "Planning Agent",
    type: "planning",
    description: "Creates detailed plans, schedules, and tracks goals",
    capabilities: ["task_planning", "scheduling", "goal_tracking", "prioritization"],
    tools: ["create_task", "get_tasks", "search_memory", "save_memory"],
    model: "claude-sonnet-4-6",
  },
  {
    id: "automation",
    name: "Automation Agent",
    type: "automation",
    description: "Automates computer tasks, browser operations, and file management",
    capabilities: ["browser_automation", "file_operations", "system_control"],
    tools: ["web_search"],
    model: "claude-sonnet-4-6",
  },
  {
    id: "memory",
    name: "Memory Agent",
    type: "memory",
    description: "Manages long-term memory, preferences, and knowledge",
    capabilities: ["memory_storage", "memory_retrieval", "context_building"],
    tools: ["search_memory", "save_memory"],
    model: "claude-haiku-4-5-20251001",
  },
  {
    id: "communication",
    name: "Communication Agent",
    type: "communication",
    description: "Drafts emails, manages calendar, handles communication",
    capabilities: ["email_drafting", "calendar_management", "meeting_scheduling"],
    tools: ["search_memory", "create_task"],
    model: "claude-sonnet-4-6",
  },
];

router.get("/", async (req: Request, res: Response) => {
  res.json(createSuccessResponse(AGENT_DEFINITIONS));
});

router.get("/:id", async (req: Request, res: Response, next) => {
  const agent = AGENT_DEFINITIONS.find((a) => a.id === req.params.id);
  if (!agent) throw new AppError(404, "NOT_FOUND", "Agent not found");
  res.json(createSuccessResponse(agent));
});

const runAgentSchema = z.object({
  goal: z.string().min(1).max(10000),
  model: z.string().optional(),
  stream: z.boolean().default(false),
});

router.post("/:id/run", chatRateLimiter, async (req: Request, res: Response, next) => {
  try {
    const agent = AGENT_DEFINITIONS.find((a) => a.id === req.params.id);
    if (!agent) throw new AppError(404, "NOT_FOUND", "Agent not found");

    const { goal, model, stream } = runAgentSchema.parse(req.body);
    const userId = req.user!.id;

    const agentRun = await prisma.agentRun.create({
      data: {
        userId,
        agentId: agent.id,
        status: "RUNNING",
        input: goal,
        startedAt: new Date(),
      },
    });

    if (stream) {
      res.setHeader("Content-Type", "text/event-stream");
      res.setHeader("Cache-Control", "no-cache");
      res.setHeader("Connection", "keep-alive");

      res.write(
        `data: ${JSON.stringify({ type: "start", runId: agentRun.id })}\n\n`
      );

      let stepNumber = 0;
      let result = "";

      result = await runAgentWithTools(userId, goal, agent.type as any, {
        model: model || agent.model,
        onStep: async (step) => {
          stepNumber++;
          await prisma.agentStep.create({
            data: {
              runId: agentRun.id,
              stepNumber,
              type: step.type.toUpperCase() as any,
              content: step.content,
              toolName: step.toolName,
            },
          });
          res.write(`data: ${JSON.stringify({ type: "step", step })}\n\n`);
        },
      });

      await prisma.agentRun.update({
        where: { id: agentRun.id },
        data: {
          status: "COMPLETE",
          output: result,
          completedAt: new Date(),
        },
      });

      res.write(
        `data: ${JSON.stringify({ type: "done", runId: agentRun.id, result })}\n\n`
      );
      res.end();
    } else {
      const result = await runAgentWithTools(userId, goal, agent.type as any, {
        model: model || agent.model,
      });

      await prisma.agentRun.update({
        where: { id: agentRun.id },
        data: { status: "COMPLETE", output: result, completedAt: new Date() },
      });

      res.json(createSuccessResponse({ runId: agentRun.id, result }));
    }
  } catch (err) {
    next(err);
  }
});

router.get("/runs/history", async (req: Request, res: Response, next) => {
  try {
    const { page = "1", pageSize = "20" } = req.query;
    const skip = (parseInt(page as string) - 1) * parseInt(pageSize as string);

    const [runs, total] = await Promise.all([
      prisma.agentRun.findMany({
        where: { userId: req.user!.id },
        orderBy: { startedAt: "desc" },
        skip,
        take: parseInt(pageSize as string),
        include: { steps: { orderBy: { stepNumber: "asc" } } },
      }),
      prisma.agentRun.count({ where: { userId: req.user!.id } }),
    ]);

    res.json(createSuccessResponse(runs, { total, page: parseInt(page as string) }));
  } catch (err) {
    next(err);
  }
});

export default router;

File: jarvis-os/services/ai/routers/agents.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional, AsyncGenerator
import json

router = APIRouter()


class AgentRunRequest(BaseModel):
    agent_type: str
    goal: str
    user_id: str
    user_context: str = ""
    model: str = "claude-sonnet-4-6"
    stream: bool = False


async def run_agent(
    agent_type: str,
    goal: str,
    user_context: str,
    model: str,
) -> str:
    if agent_type == "research":
        from agents.research_agent import run_research_agent
        return await run_research_agent(goal, user_context, model)

    # Fallback for other agent types
    from agents.research_agent import get_llm
    from langchain_core.messages import HumanMessage, SystemMessage

    AGENT_PROMPTS = {
        "coding": "You are JARVIS's Coding Agent. Generate, analyze, and explain code with precision.",
        "planning": "You are JARVIS's Planning Agent. Create detailed, actionable plans.",
        "automation": "You are JARVIS's Automation Agent. Plan computer and browser automation tasks.",
        "memory": "You are JARVIS's Memory Agent. Organize and retrieve information efficiently.",
        "communication": "You are JARVIS's Communication Agent. Draft professional communications.",
    }

    system = AGENT_PROMPTS.get(
        agent_type,
        "You are JARVIS, an advanced AI assistant. Complete the given goal."
    )

    llm = get_llm(model)
    messages = [
        SystemMessage(content=f"{system}\n\n{user_context}"),
        HumanMessage(content=goal),
    ]
    response = await llm.ainvoke(messages)
    return response.content


async def stream_agent(
    agent_type: str,
    goal: str,
    user_context: str,
    model: str,
) -> AsyncGenerator[str, None]:
    if agent_type == "research":
        from agents.research_agent import stream_research_agent
        async for chunk in stream_research_agent(goal, user_context, model):
            yield f"data: {json.dumps({'type': 'delta', 'content': chunk})}\n\n"
        yield f"data: {json.dumps({'type': 'done'})}\n\n"
        return

    from agents.research_agent import get_llm
    from langchain_core.messages import HumanMessage, SystemMessage

    llm = get_llm(model)
    messages = [
        SystemMessage(content=f"You are JARVIS's {agent_type} agent.\n\n{user_context}"),
        HumanMessage(content=goal),
    ]

    async for chunk in llm.astream(messages):
        if chunk.content:
            yield f"data: {json.dumps({'type': 'delta', 'content': chunk.content})}\n\n"

    yield f"data: {json.dumps({'type': 'done'})}\n\n"


@router.post("/run")
async def run_agent_endpoint(request: AgentRunRequest):
    if request.stream:
        return StreamingResponse(
            stream_agent(
                request.agent_type,
                request.goal,
                request.user_context,
                request.model,
            ),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
            },
        )

    result = await run_agent(
        request.agent_type,
        request.goal,
        request.user_context,
        request.model,
    )
    return {"success": True, "data": {"result": result}}


@router.get("/types")
async def list_agent_types():
    return {
        "success": True,
        "data": [
            {
                "type": "research",
                "name": "Research Agent",
                "description": "Web search and information gathering",
            },
            {
                "type": "coding",
                "name": "Coding Agent",
                "description": "Code generation, analysis, debugging",
            },
            {
                "type": "planning",
                "name": "Planning Agent",
                "description": "Task planning and goal tracking",
            },
            {
                "type": "automation",
                "name": "Automation Agent",
                "description": "Computer and browser automation",
            },
            {
                "type": "memory",
                "name": "Memory Agent",
                "description": "Long-term memory management",
            },
            {
                "type": "communication",
                "name": "Communication Agent",
                "description": "Email and calendar management",
            },
        ],
    }
